### DISCUSSION TITLE
What's the suggested resolution for Assert.ThrowsAsync compiler ambiguity with 2.6.0 and async anonymous functions?

### DISCUSSION THREAD

**Discussion by jskeet** [xunit/xunit#2807] [Question]
Consider the following test, targeting net6.0:

```csharp
public class UnitTest1
{
    [Fact]
    public async Task RangeCheck()
    {
        await Assert.ThrowsAsync<ArgumentOutOfRangeException>(async () => await CheckAsync(4, 3));
    }

    private async Task CheckAsync(int min, int max)
    {
        if (min > max)
        {
            throw new ArgumentOutOfRangeException();
        }
        // Whatever the actual code would do
        await Task.Yield();
    }
}
```

This works fine in 2.5.3; in 2.6.0 it fails because the `ThrowsAsync` call is ambiguous between `ThrowsAsync(Func<Task>)` and `ThrowsAsync(Func<ValueTask>)`. The [release notes](https://xunit.net/releases/v2/2.6.0) allude to this, but refer to [issue 2793](https://github.com/xunit/xunit/issues/2793) which is about different ambiguous calls.

In my actual code, in some cases I've managed to remove the ambiguity by changing from an anonymous function to a method group conversion - but that doesn't work in the case above where I need to pass arguments.

Declaring a local variable first would work:

```csharp
Func<Task> work = async () => await CheckAsync(4, 3);
await Assert.ThrowsAsync<ArgumentOutOfRangeException>(work);
```

... and I suspect with later versions of C# it could even be implicitly typed... but it's a bit ugly.

Any suggestions for a more elegant approach? (We might want to add to issue 2793 for this, too.)

(Slightly later...)

Thinking about it more, for a simple call like this, we could just use:

```csharp
await Assert.ThrowsAsync<ArgumentOutOfRangeException>(() => CheckAsync(4, 3));
```

That doesn't help when you need to await within the anonymous function though. That's only relevant for more complex scenarios, and maybe it's okay just to extract the local variable for those cases...

**Comment by bradwilson** [comment]
Raised in issue https://github.com/xunit/xunit/issues/2808, please add additional commentary there.

I've opened an issue with the C# compiler to see if they can resolve this. I'd rather not rename functions which would reduce visibility but I may end up having to do that.
